js/Vector.h produces dizzying number of "field 'storage' is uninitialized when used" warnings

```
14:07.35 Warning: -Wuninitialized in /Users/pale/jenkins/workspace/palemoon/obj-x86_64-apple-darwin12.5.0/dist/include/js/Vector.h: field 'storage' is uninitialized when used here
14:07.35 ./../../dist/include/js/Vector.h:531:34: warning: field 'storage' is uninitialized when used here [-Wuninitialized]
14:07.37   : AllocPolicy(ap), mBegin((T *)storage.addr()), mLength(0),
14:07.37                                  ^
14:07.38 /Users/pale/jenkins/workspace/palemoon/js/src/jsgc.h:705:5: note: in instantiation of member function 'js::Vector<void **, 16, js::SystemAllocPolicy>::Vector' requested here
14:07.38     GCHelperThread(JSRuntime *rt)
```

A huge number (at least hundreds, maybe thousands?) of source files produce warnings much like the above -- I'm not sure whether it's harmful or not, but it seems like fixing it once in the header file could dramatically reduce the size of the build output (and probably therefore the speed of the build, as logging to stdout is a measurable amount of time in such a large build)

This is with:
```
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 5.1 (clang-503.0.40) (based on LLVM 3.4svn)
Target: x86_64-apple-darwin12.5.0
Thread model: posix
```

--- a/js/public/Vector.h
+++ b/js/public/Vector.h
@@ -525,6 +525,11 @@ class Vector : private AllocPolicy
 
 /* Vector Implementation */
 
+// Clang unnecessarily warns 'uninitialized use of storage', and since
+// this is a header it warns for a huge number of files... #298
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wuninitialized"
+
 template <class T, size_t N, class AllocPolicy>
 JS_ALWAYS_INLINE
 Vector<T,N,AllocPolicy>::Vector(AllocPolicy ap)
@@ -534,6 +539,8 @@ Vector<T,N,AllocPolicy>::Vector(AllocPolicy ap)
   , mReserved(sInlineCapacity), entered(false)
 #endif
 {}
+
+#pragma clang diagnostic pop
 
 /* Move constructor. */
 template <class T, size_t N, class AllocPolicy>

